[BUG] Rector removes class in use

# Bug Report

| Subject        | Details                                                         |
| :------------- | :---------------------------------------------------------------|
| Rector version | 0.9.16              |
| Installed as   | composer dependency                           |
| PHP| 7.4.14 | 

Regarding version:
My `composer.json` says `0.9.16` and `rector --version` returns 
```shell 
vendor/bin/rector --version                                     
Rector Unknown
``` 
<!-- Please describe your problem here. -->

I have build a small Laravel zero application to reproduce this problem.

When running the `vendor/bin/rector process --dry-run --config rector-ci.php` it wants to remove the `Models/Vehicle.php` file, and I don't know why. 

The `Car`-object is used in the `InspiringCommand.php` and calls the inherited method `hasEngine()` therefore I don't see why rector don't recognize this as in use. The `extend Vehicle` itself should in my eyes be enough for that and not want to remove it.  

```shell 
[WARNING] File "app/Models/Vehicle.php" will be removed  
```

If I change the class to be and abstract class it will stay. But not every class that I might want to inherit will be abstract. 

## Minimal PHP Code Causing Issue

I have stripped away all not needed. My file tree looks like this for the files I'll list below
```shell
app
├── Commands
│   └── InspiringCommand.php
├── Models
│   ├── Car.php
│   └── Vehicle.php
``` 
`Commands/InspiringCommand.php`
```php
<?php

namespace App\Commands;
use App\Models\Car;

final class InspiringCommand
{
    protected $signature = 'inspiring {name=Artisan}';
    protected $description = 'Display an inspiring quote';
    public function handle(): void
    {
        $car = new Car();
        $this->info($car->hasEngine());
        $this->info('Simplicity is the ultimate sophistication.');
    }
}
``` 
`Models/Car.php`
```php
<?php

declare(strict_types=1);
namespace App\Models;

class Car extends Vehicle
{
    private string $engineSize;
}
``` 

`Models/Vehicle.php`
```php
<?php

declare(strict_types=1);
namespace App\Models;

class Vehicle
{
    protected string $colour;
    protected bool $engine;

    public function hasEngine(): bool
    {
        return $this->engine;
    }
}
```
`rector-ci.php` The rules are adjusted to this example to only have the errors that troubles me atm. 
```php
<?php

declare(strict_types=1);

use Rector\Core\Configuration\Option;
use Rector\DeadCode\Rector\Cast\RecastingRemovalRector;
use Rector\DeadCode\Rector\Property\RemoveUnusedPrivatePropertyRector;
use Rector\Set\ValueObject\SetList;
use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;

return static function (ContainerConfigurator $containerConfigurator): void {
    $parameters = $containerConfigurator->parameters();
    $parameters->set(Option::PATHS, [
            __DIR__ . '/app',
            __DIR__ . '/tests',
            __DIR__ . '/rector-ci.php'
        ]
    );
    $parameters->set(Option::SETS, [
            SetList::CODE_QUALITY,
            SetList::CODE_QUALITY_STRICT,
            SetList::CODING_STYLE,
            SetList::DEAD_CLASSES,
            SetList::DEAD_CODE,
            SetList::DEAD_DOC_BLOCK,
            SetList::DEFLUENT,
            SetList::EARLY_RETURN,
            SetList::PHP_74,
            SetList::PHPUNIT_CODE_QUALITY,
            SetList::PRIVATIZATION,
            SetList::NAMING,
            SetList::TYPE_DECLARATION,
            SetList::LARAVEL_CODE_QUALITY,
            SetList::PHP_CODE_SNIFFER_30,
        ]
    );
    $parameters->set(Option::AUTO_IMPORT_NAMES, false);
    $parameters->set(Option::AUTOLOAD_PATHS, [__DIR__ . '/app']);
    $parameters->set(Option::SKIP,
        [
            Rector\Privatization\Rector\ClassMethod\PrivatizeLocalOnlyMethodRector::class,
            Rector\DeadCode\Rector\ClassMethod\RemoveUnusedPrivateMethodRector::class,
            Rector\DeadCode\Rector\ClassMethod\RemoveEmptyClassMethodRector::class,
        ],
    );

    $services = $containerConfigurator->services();
    $services->set(RemoveUnusedPrivatePropertyRector::class);
};

```
As it's regarding removing files I didn't find a way to reproduce this with the Rector Demo tool. 

## Expected Behaviour

I would expect the `Vehicle.php` to not be removed as it's clearly in use.

I'm trying to add tests now. But the `composer complete-check` gives me following error, and I didn't do any changes yet. 

Only did a 

```shell
git clone
composer install 
composer complete-check
``` 

```shell
vendor/bin/phpstan analyse --ansi --error-format symplify
Note: Using configuration file /home/tomas/Projects/php/rector/phpstan.neon.
 2624/2624 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

 ---------------------------------------------------------------------------------------------------
  rules/composer/src/Rector/AddPackageToRequireComposerRector.php:71
 ---------------------------------------------------------------------------------------------------
  - '#Content of method "configure\(\)" is duplicated with method "configure\(\)" in "Rector\\Composer\\Rector\\ChangePackageVersionComposerRector" class\. Use unique content or abstract service instead#'
 ---------------------------------------------------------------------------------------------------

 ---------------------------------------------------------------------------------------------------
  src/Util/PhpVersionFactory.php:9
 ---------------------------------------------------------------------------------------------------
  - '#Class with base "PhpVersionFactory" name is already used in "PHPStan\\Php\\PhpVersionFactory", "Rector\\Core\\Util\\PhpVersionFactory"\. Use unique name to make classes easy to recognize#'
 ---------------------------------------------------------------------------------------------------


                                                                                                           
 [ERROR] Found 2 errors                                                                                    
                                                                                                           

Script vendor/bin/phpstan analyse --ansi --error-format symplify handling the phpstan event returned with error code 1
Script @phpstan was called via complete-check
```

--- a/rules/dead-code/src/Rector/Class_/RemoveUnusedClassesRector.php
+++ b/rules/dead-code/src/Rector/Class_/RemoveUnusedClassesRector.php
@@ -140,7 +140,11 @@ CODE_SAMPLE
             return true;
         }
 
-        return $class->isAbstract();
+        if ($class->isAbstract()) {
+            return true;
+        }
+
+        return $this->nodeRepository->hasClassChildren($class);
     }
 
     private function hasMethodWithApiAnnotation(Class_ $class): bool

--- a/rules/dead-code/tests/Rector/Class_/RemoveUnusedClassesRector/Fixture/skip_class_extend.php.inc
+++ b/rules/dead-code/tests/Rector/Class_/RemoveUnusedClassesRector/Fixture/skip_class_extend.php.inc
@@ -2,7 +2,7 @@
 
 namespace Rector\DeadCode\Tests\Rector\Class_\RemoveUnusedClassesRector\Fixture;
 
-class Car extends Vehicle
+class SkipClassExtend extends Vehicle
 {
     /**
      * @var string

--- a/rules/dead-code/tests/Rector/Class_/RemoveUnusedClassesRector/Fixture/skip_class_extend.php.inc
+++ b/rules/dead-code/tests/Rector/Class_/RemoveUnusedClassesRector/Fixture/skip_class_extend.php.inc
@@ -2,20 +2,17 @@
 
 namespace Rector\DeadCode\Tests\Rector\Class_\RemoveUnusedClassesRector\Fixture;
 
-class skipClassExtend
-{
-    public function run(): void
-    {
-        echo (new Car())->hasEngine();
-    }
-}
-
 class Car extends Vehicle
 {
     /**
      * @var string
      */
     private $engineSize;
+
+    public function run(): void
+    {
+        echo (new self())->hasEngine();
+    }
 }
 
 class Vehicle

--- a/rules/dead-code/tests/Rector/Class_/RemoveUnusedClassesRector/Fixture/skip_class_extend.php.inc
+++ b/rules/dead-code/tests/Rector/Class_/RemoveUnusedClassesRector/Fixture/skip_class_extend.php.inc
@@ -1,0 +1,37 @@
+<?php
+
+namespace Rector\DeadCode\Tests\Rector\Class_\RemoveUnusedClassesRector\Fixture;
+
+class skipClassExtend
+{
+    public function run(): void
+    {
+        echo (new Car())->hasEngine();
+    }
+}
+
+class Car extends Vehicle
+{
+    /**
+     * @var string
+     */
+    private $engineSize;
+}
+
+class Vehicle
+{
+    /**
+     * @var string
+     */
+    protected $colour;
+
+    /**
+     * @var bool
+     */
+    protected $engine;
+
+    public function hasEngine(): bool
+    {
+        return $this->engine;
+    }
+}

